### PR TITLE
Functionality to track "first run" in cache objects

### PR DESCRIPTION
**Pull request by StanleyGoldman**
As mentioned in #409

We want to start using the Invalidated events on cache objects to know when we need to re-populate data.

For some of those cache objects, we want to fire an implicit invalidated on the application's "First Run" as opposed to a domain reload. That way we know these objects need to be populated immediately.

This change adds:
- functionality to track the first run time in `ApplicationCache`
- functionality to track when `ManagedCacheBase` objects are initialized
- functionality in `ManagedCacheBase.ValidateData()` to verify if cache objects were initialized after application first run time, otherwise it will update and fire an "Invalidated" event
- functionality to define which cache objects are invalid on first tun

This provides base functionality to:
- #427 
- #436 

Depends on:
- [x] #435

### CODE REVIEW

File: src/UnityExtension/Assets/Editor/GitHub.Unity/ApplicationCache.cs [hunk 1]
```diff
@@ -178,6 +213,24 @@ public DateTimeOffset LastVerifiedAt
...
+                if (!initializedAtValue.HasValue)
+                {
+                    initializedAtValue = DateTimeOffset.Parse(InitializedAtString);
+                }
+
+                return initializedAtValue.Value;
+            }
+            set
+            {
+                InitializedAtString = value.ToString();
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/ApplicationCache.cs]: All of these `DateTimeOffset.ToString()` calls need to specify the ISO 8016 format, eg. `2017-11-16T20:24:46+00:00`. As it is, they're outputting strings in whatever local configuration the system has, which may not be parseable by `Parse`.

Comment (StanleyGoldman) [reply src/UnityExtension/Assets/Editor/GitHub.Unity/ApplicationCache.cs]: Originally fixed in #457 carried over here

**Comment by StanleyGoldman** [comment]
@jcansdale Unity cannot serialize `DateTimeOffset` objects, but it will serialize `string` objects. So for all these dates we have a `string` field and a `DateTimeOffset?` field; the latter of which we populate lazily. I did however make a change based on your comment, there was no reason to set the lazily generated value to null and thus wait for the lazy generation.

**Comment by sguthals** [comment]
Seems straight forward and like a clean approach.

**Pull request merged**